### PR TITLE
Add support for luks resource

### LINKED ISSUE #14: Feature Request: `luks`
The Ignition spec now ([3.2.0](https://coreos.github.io/ignition/configuration-v3_2/)) includes a `luks` section which will need a data source added to this provider.

The interface might look something like:
```terraform
data "ignition_luks" "example_disk" {
  name = "example_disk"
  device = "/dev/disk/by-uuid/d6380b03-38ff-4c16-8dbd-3ee864cc3fbf"
  clevis {
    tpm2 = true
  }
}
```

### PR DESCRIPTION
**Pull request by n0rad**
closes #14

### CODE REVIEW

File: ignition/resource_ignition_luks.go [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+							Type:     schema.TypeBool,
+							ForceNew: true,
+							Optional: true,
+						},
+						"threshold": {
+							Type:     schema.TypeInt,
+							ForceNew: true,
+							Optional: true,
+						},
+						"custom": {
```
Comment (yussufsh) [inline ignition/resource_ignition_luks.go:56]: Please add `MaxItems: 1,`

File: ignition/resource_ignition_luks.go [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+						},
+					},
+				},
+			},
+			"device": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"key_file": {
```
Comment (yussufsh) [inline ignition/resource_ignition_luks.go:94]: Please add `MaxItems: 1,`

File: ignition/resource_ignition_luks.go [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+	if _, hasKeyFile := d.GetOk("key_file"); hasKeyFile {
+		var keyFile types.Resource
+		if err := fillResource(d, &keyFile, "key_file"); err != nil {
+			return "", err
+		}
+		luks.KeyFile = keyFile
+	}
+
+	if discard, hasDiscard := d.GetOk("discard"); hasDiscard {
```
Comment (yussufsh) [inline ignition/resource_ignition_luks.go:207]: Am I missing something? no `discard` defined in the schema or docs.

File: website/docs/d/luks.html.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+* `clevis` - (Optional) Describes the clevis configuration for the luks device.
+
+The `key_file` block supports:
+
+* `source` - (Required) The URL of the key file. Supported schemes are http, https, tftp, s3, arn, gs, and data. When using http, it is advisable to use the verification option to ensure the contents haven’t been modified.
+
+* `compression` - (Optional) The type of compression used on the key file (null or gzip). Compression cannot be used with S3.
+
+* `httpHeaders` - (Optional) A list of HTTP headers to be added to the request. Available for http and https source schemes only.
```
Comment (yussufsh) [inline website/docs/d/luks.html.md]:
```suggestion
* `http_headers` - (Optional) A list of HTTP headers to be added to the request. Available for http and https source schemes only.
```

File: website/docs/d/luks.html.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+* `verification` - (Optional) Options related to the verification of the key file.
+
+The `verification` block supports:
+
+* `hash` - (Optional) The hash of the key file, in the form <type>-<value> where type is either sha512 or sha256. If compression is specified, the hash describes the decompressed key file.
```
Comment (yussufsh) [inline website/docs/d/luks.html.md]: There is no `verification.hash` defined in the schema. Please move the same description to `verification`.

File: website/docs/d/luks.html.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+* `advertisement` - (Optional) The advertisement JSON. If not specified, the advertisement is fetched from the tang server during provisioning.
+
+The `custom` block supports:
+
+* `pin` - (Optional) The clevis pin.
+
+* `config` - (Optional) The clevis configuration JSON.
+
+* `needsNetwork` - (Optional) Whether or not the device requires networking.
```
Comment (yussufsh) [inline website/docs/d/luks.html.md]:
```suggestion
* `needs_network` - (Optional) Whether or not the device requires networking.
```

File: website/docs/d/luks.html.md [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+The `key_file` block supports:
+
+* `source` - (Required) The URL of the key file. Supported schemes are http, https, tftp, s3, arn, gs, and data. When using http, it is advisable to use the verification option to ensure the contents haven’t been modified.
+
+* `compression` - (Optional) The type of compression used on the key file (null or gzip). Compression cannot be used with S3.
+
+* `httpHeaders` - (Optional) A list of HTTP headers to be added to the request. Available for http and https source schemes only.
+
+* `verification` - (Optional) Options related to the verification of the key file.
+
```
Comment (yussufsh) [inline website/docs/d/luks.html.md:56]: Please add that `The http_headers blocks support:`...

**Review by yussufsh — CHANGES_REQUESTED** [review]

File: ignition/resource_ignition_luks.go (hunk 3 above)
Comment (n0rad) [reply ignition/resource_ignition_luks.go:207]: https://github.com/coreos/ignition/blob/main/config/v3_4/types/schema.go#L114

File: website/docs/d/luks.html.md (hunk 5 above)
Comment (n0rad) [reply website/docs/d/luks.html.md]: https://github.com/coreos/ignition/blob/main/config/v3_4/types/schema.go#L258

Comment (yussufsh) [reply website/docs/d/luks.html.md]: We do not have `hash` in our resource schema definition. See https://github.com/community-terraform-providers/terraform-provider-ignition/blob/v2/ignition/resource_ignition_config.go#L41-L45

Example doc: https://registry.terraform.io/providers/community-terraform-providers/ignition/latest/docs/data-sources/file#verification

File: ignition/resource_ignition_luks.go (hunk 3 above)
Comment (yussufsh) [reply ignition/resource_ignition_luks.go:207]: I mean it is not defined anywhere in Resource Schema above.

**Review by yussufsh — CHANGES_REQUESTED** [review]

File: website/docs/d/luks.html.md (hunk 5 above)
Comment (n0rad) [reply website/docs/d/luks.html.md]: ah! good catch, let me fix that

File: website/docs/d/luks.html.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+* `http_headers` - (Optional) A list of HTTP headers to be added to the request. Available for http and https source schemes only.
+
+* `verification` - (Optional) The hash of the config, in the form _\<type\>-\<value\>_ where type is either sha512 or sha256. If compression is specified, the hash describes the decompressed file.
+
+The `http_headers` block supports:
+
+* `name` - (Required) The header name.
+
+* `value` - (Optional) The header contents.
```
Comment (yussufsh) [inline website/docs/d/luks.html.md]:
```suggestion
* `value` - (Required) The header contents.
```

File: website/docs/d/luks.html.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+* `tpm2` - (Optional) Whether or not to use a tpm2 device.
+
+* `threshold` - (Optional) Sets the minimum number of pieces required to decrypt the device. Default is 1.
+
+* `custom` - (Optional) Overrides the clevis configuration. The pin & config will be passed directly to clevis luks bind. If specified, all other clevis options must be omitted.
+
+The `tang` block supports:
+
+* `url` - (Optional) Url of the tang server.
```
Comment (yussufsh) [inline website/docs/d/luks.html.md]:
```suggestion
* `url` - (Required) Url of the tang server.
```

**Review by yussufsh — CHANGES_REQUESTED** [review]

File: website/docs/d/luks.html.md (hunk 8 above)
Comment (n0rad) [reply website/docs/d/luks.html.md]: Note that the `value` of headers is optional in ignition, while declared as required in terraform's `httpHeaderReferenceResource`.

**Review by yussufsh — APPROVED** [review]
lgtm

**Pull request merged**